scripts/cddl_gen.py

```diff
@@ -0,0 +1,1267 @@
+            raise ValueError("Did not expect multiple parsed values for union")
+
+        self.type = type
+
+        value = valueGenerator()
+        self.value = value
+
+        if type in ["BSTR", "TSTR"]:
+            if value is not None:
+                self.setSize(len(value))
+        if type in ["UINT", "NINT"]:
+            if value is not None:
+                self.size = sizeof(value)
+                self.minValue = value
+                self.maxValue = value
+
+
+    # Set the self.type and self.minValue and self.maxValue (or self.minSize and self.maxSize depending on the type) of
+    # this element. For use during CDDL parsing.
+    def typeAndRange(self, type, minVal, maxVal):
+        if type not in ["INT", "UINT", "NINT"]:
+            raise TypeError("Only integers (not %s) can have range" % (type,))
+        if minVal > maxVal:
+            raise TypeError("Range has larger minimum than maximum (min %d, max %d)" % (minVal, maxVal))
+        if minVal == maxVal:
+            return self.typeAndValue(type, minVal)
+        self.type = type
+        self.minValue = minVal
+        self.maxValue = maxVal
+        if type in "UINT":
+            self.setSizeRange(sizeof(minVal), sizeof(maxVal))
+        if type is "NINT":
+            self.setSizeRange(sizeof(abs(maxVal)), sizeof(abs(minVal)))
+        if type is "INT":
+            self.setSizeRange(0, max(sizeof(abs(maxVal)), sizeof(abs(minVal))))
+
+
+    # Set the self.value and self.size of this element. For use during CDDL parsing.
+    def typeValueSize(self, type, value, size):
+        self.typeAndValue(type, value)
+        self.setSize(size)
+
+    # Set the self.label of this element. For use during CDDL parsing.
+    def setLabel(self, label):
+        if self.type is not None: raise TypeError("Cannot have label after value: " + label)
+        self.label = label
+
+    # Set the self.quantifier, self.minQ, and self.maxQ of this element. For use during CDDL parsing.
+    def setQuantifier(self, quantifier):
+        if self.type is not None: raise TypeError("Cannot have quantifier after value: " + quantifier)
+
+        quantifierMapping =\
+        [
+            ("\?", lambda matchObj: (0, 1)),
+            ("\*", lambda matchObj: (0, None)),
+            ("\+", lambda matchObj: (1, None)),
+            ("(\d*)\*(\d*)", lambda matchObj: (int(matchObj.groups()[0]), int(matchObj.groups()[1]))),
+        ]
+
+        self.quantifier = quantifier
+        for (reg, handler) in quantifierMapping:
+            matchObj = match(reg, quantifier)
+            if matchObj:
+                (self.minQ, self.maxQ) = handler(matchObj)
+                if self.maxQ is None: self.maxQ = self.DEFAULT_MAXQ
+                return
+        raise ValueError("invalid quantifier: %s" % quantifier)
+
+    # Set the self.size of this element. This will also set the self.minValue and self.maxValue of UINT types.
+    # For use during CDDL parsing.
+    def setSize(self, size):
```

Comment:
The functions setSize and SetSizeRange is callilng each other..  Potentially makes understanding flow harder than it needs to be. Could this be rewritten to a single function with the use of named parameters (defaulted to None)

This would also make the handling invalid types in an entry assert-statement, clearer (Type: None, UINT, BSTR and TSTR)

Reply:
Fixed